presto-main-base/src/main/java/com/facebook/presto/sql/rewrite/NativeExecutionTypeRewrite.java

```diff
@@ -121,28 +125,74 @@ protected Node visitCast(Cast node, Void context)
             catch (IllegalArgumentException | UnknownTypeException e) {
                 throw new SemanticException(TYPE_MISMATCH, node, "Unknown type: " + node.getType());
             }
-            return node;
+
+            Expression expression = node.getExpression();
+            expression = peelIfDereferenceExpression(expression);
+            return new Cast(node.getLocation(), expression, node.getType(), node.isSafe(), node.isTypeOnly());
         }
 
         @Override
         protected Node visitFunctionCall(FunctionCall node, Void context)
         {
-            if (isValidEnumKeyFunctionCall(node)) {
-                Cast argument = (Cast) node.getArguments().get(0);
-                Type argumentType = functionAndTypeResolver.getType(parseTypeSignature(argument.getType()));
-                if (argumentType instanceof TypeWithName) {
-                    // Peel user defined type name.
-                    argumentType = ((TypeWithName) argumentType).getType();
+            QualifiedName functionName = node.getName();
+            List<Expression> arguments = node.getArguments();
+            if (node.getArguments().size() == 1) {
+                Expression argument = arguments.get(0);
+                if (isValidEnumKeyFunctionCall(node)) {
+                    functionName = QualifiedName.of(FUNCTION_ELEMENT_AT);
+                    Type argumentType;
+                    if (argument instanceof Cast) {
+                        argumentType = functionAndTypeResolver.getType(parseTypeSignature(((Cast) argument).getType()));
+                    }
+                    else if (argument instanceof DereferenceExpression) {
+                        argumentType = functionAndTypeResolver.getType(parseTypeSignature(((DereferenceExpression) argument).getBase().toString()));
+                    }
+                    else {
+                        // ENUM_KEY is only supported with Cast or DereferenceExpression for now.
+                        // Return node without rewriting.
+                        return node;
+                    }
+                    argument = rewriteIfCastOrDereferenceExpression(argument);
```

Comment:
instead of this custom rewriteIfCastOrdereferenceExpression, why don't we visit all the argument s